core/emulate.c

```diff
@@ -1169,8 +1173,8 @@ em_status_t EMCALL em_emulate_insn(struct em_context_t *ctxt)
             decode_operands(ctxt);
             goto restart;
         }
-        if ((ctxt->rep == PREFIX_REPNE && (ctxt->rflags & RFLAGS_ZF)) ||
-            (ctxt->rep == PREFIX_REPE && !(ctxt->rflags & RFLAGS_ZF))) {
+        if ((ctxt->rep == PREFIX_REPNE && !(ctxt->rflags & RFLAGS_ZF)) ||
```

Comment:
nit: Good catch! I think it's worth mentioning in the cover page, and maybe even moving into a separate commit.

With this change, we're closer to the spec, but still not 100% compliant. Here are the condition checks given in the pseudocode for REP in Intel SDM (Vol. 2B, Section 4.3: _Instructions (M-U)_, _REP/REPE/REPZ/REPNE/REPNZ—Repeat String Operation Prefix_):

```
WHILE CountReg ≠ 0
  DO
      Service pending interrupts (if any);
      Execute associated string instruction;
      CountReg ← (CountReg – 1);
      IF CountReg = 0
          THEN exit WHILE loop; FI;
      IF (Repeat prefix is REPZ or REPE) and (ZF = 0)
      or (Repeat prefix is REPNZ or REPNE) and (ZF = 1)
          THEN exit WHILE loop; FI;
  OD;
```

Note that it checks RCX right after decrementing it. If RCX == 0, probably we should skip `decode_operands()`? If you agree, please implement it in a separate commit (which can be combined with the `ZF` condition fix).

Reply:
I've fixed this in 3ee3b48. Note that `decode_operands` does not access memory: it's called just to update the `ea` field for `op_di` and `op_si` operands, so that the correct memory address is read later on. Then after `goto restart` we just check whether RCX != 0, so I'd say the current implementation complies with the specification (unless I'm missing something else?).

Nonetheless I've moved the `decode_operands` stage right after the RCX != 0 check.

Reply:
Ah, you're right! I just blindly assumed that `decode_operands` would access memory, sorry. But the new patch does look cleaner :-)